src/complete.cpp

```diff
@@ -224,13 +224,13 @@ completion_t &completion_t::operator=(const completion_t &him) {
 }
 
 bool completion_t::is_naturally_less_than(const completion_t &a, const completion_t &b) {
+    //for this to work, stable_sort must be used because results aren't interchangeable
+    if ((a.flags & COMPLETE_DONT_SORT) != 0 && (b.flags & COMPLETE_DONT_SORT) != 0) {
```

Comment:
Yeah, you're right. I was only considering the common case where the completions are all from a single homogenous source. So that `if()` needs to stay as is.

As for the `!= 0` test I would omit it. The `var & FLAG_ENUM` idiom for testing if a particular flag is set should be clear to even a relatively new C programmer. Because the code has been touched by a lot of people over more than a decade you'll find the project isn't very consistent about such things. But my previous job was at Google where maintaining a consistent style is only slightly less important than correctness and speed of the code. So I'm slowly chipping away at reforming fish code to have a consistent style. For example, I implemented a `make style` build target to automatically run changes through `clang-format`.